PA193_mnemonic_Slytherin/mnemonic.py

```diff
@@ -11,159 +11,241 @@
 
+        words = [self.__dict_list[i] for i in indexes]
+        return Mnemonic(' '.join(words))
 
-def _secure_seed_compare(expected_seed: bytes, actual_seed: bytes) -> bool:
-    """Compare provided seeds in constant time to prevent timing attacks.
-    :raises TypeError: if parameters are not bytes-like objects
-    :rtype: bool
-    :return: True if seeds are the same, False otherwise.
+
+class Mnemonic(str, dictionaryAccess):
+    """Class for mnemonic representation.
     """
-    # > hmac.compare_digest` uses an approach designed to prevent timing
-    # > analysis by avoiding content-based short circuiting behaviour, making
-    #  > it appropriate for cryptography
-    # > https://docs.python.org/3.7/library/hmac.html#hmac.compare_digest
-    #
-    # > Note: If a and b are of different lengths, or if an error occurs,
-    # > a timing attack could theoretically reveal information about the types
-    #  > and lengths of a and b—but not their values.
-    #
-    # Type and length of seeds is known to the attacker, but not the value of expected seed.
-    if not (isinstance(expected_seed, bytes) and isinstance(actual_seed, bytes)):
-        # Function `hmac.compare_digest` accepts strings & bytes and raises TypeError
-        # for other types. It would accept two strings, but we accept only two
-        # bytes-like objects, therefore we raise TypeError here.
-        raise TypeError('a bytes-like object is required')
-    return hmac.compare_digest(expected_seed, actual_seed)
-
-
-def generate(entropy: bytes, seed_password: str = '') -> Tuple[str, bytes]:
+    
+    def __init__(self, mnemonic: str):
+        """Convert mnemonic phrase to entropy using dictionary to ensure its validity.
+        :raises ValueError: on invalid parameters
+        """
+        if not isinstance(mnemonic, str):
+            raise ValueError('Cannot instantiate mnemonic')
+    
+        words = mnemonic.split()
+        l = len(words)
+        if not l in (12, 15, 18, 21, 24):
+            raise ValueError('Cannot instantiate mnemonic')
+
+        try:
+            indexes = [self.__dict_dict[word] for word in words]
+        except KeyError:
+            raise ValueError('Cannot instantiate mnemonic')
+
+        # Concatenate indexes into single 
+        indexes_bin = sum([indexes[-i - 1] << i * 11 for i in reversed(range(l))])
+
+        # Number of bits entropy is shifted by
+        shift = l * 11 // 32
+
+        checksum = indexes_bin & (pow(2, shift) - 1)
+        entropy_bin =  indexes_bin >> shift
+        entropy = entropy_bin.to_bytes((l * 11 - shift) // 8, byteorder='big')
+    
+        self.__entropy = Entropy(entropy)
+        # Check correctness
+        if checksum != self.__entropy.checksum(shift):
+            raise ValueError('Cannot instantiate mnemonic')
+        
+        self.__repr__ = mnemonic
+
+
+    def toSeed(self, seed_password: str = '') -> Seed:
```

Comment:
I wonder if we need another class for password (or even a passphrase itself), since no one checks the length of the password at the moment.

Reply:
Don't think so. It's just normal string which you check as parameter of method where it's used. There's no other processing of password or internal state of password.